apollo-runtime/src/jsMain/kotlin/com/apollographql/apollo3/network/ws/JsWebSocketEngine.kt

```diff
@@ -0,0 +1,117 @@
+import io.ktor.http.Url
+import io.ktor.util.PlatformUtils
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.suspendCancellableCoroutine
+import okio.ByteString
+import org.khronos.webgl.Uint8Array
+import org.w3c.dom.MessageEvent
+import org.w3c.dom.WebSocket
+import org.w3c.dom.events.Event
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+
+actual class DefaultWebSocketEngine : WebSocketEngine {
+
+  override suspend fun open(
+      url: String,
+      headers: Map<String, String>,
+  ): WebSocketConnection = open(Url(url), headers)
+
+  private suspend fun open(url: Url, headers: Map<String, String>): WebSocketConnection {
+    val newUrl = url.copy(
+        protocol = when (url.protocol) {
+          URLProtocol.HTTPS -> URLProtocol.WSS
+          URLProtocol.HTTP -> URLProtocol.WS
+          URLProtocol.WS, URLProtocol.WSS -> url.protocol
+          /* URLProtocol.SOCKS */else -> throw UnsupportedOperationException("SOCKS is not a supported protocol")
+      })
+    val socket = createWebSocket(newUrl.toString(), Headers.build { headers.forEach { append(it.key, it.value) } }).awaitConnection()
+    val messageChannel = ChannelWrapper(Channel<String>(Channel.UNLIMITED))
+    socket.onmessage = { messageEvent: MessageEvent ->
+      val data = messageEvent.data
+      if (data != null) {
+        when (data) {
+          is String -> messageChannel.trySend(data)
+          else -> throw UnsupportedOperationException("${data::class.simpleName} is currently unsupported")
+        }
+      }
+    }
+    return object : WebSocketConnection {
+      override suspend fun receive(): String {
+        return messageChannel.receive()
+      }
+
+      override fun send(data: ByteString) {
+        socket.send(Uint8Array(data.toByteArray().toTypedArray()))
+      }
+
+      override fun send(string: String) {
+        socket.send(string)
+      }
+
+      override fun close() {
+        socket.close(CLOSE_NORMAL.toShort())
+        //close the message channel as well, since it is idempotent there is no harm
+        messageChannel.close()
+      }
+
+    }
+  }
+
+  /*
+   * The following applies for lines 71-114
+   * Copyright 2014-2019 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+   * Some lines have been added/modified from the original source (https://github.com/ktorio/ktor/blob/f723638afd4d36024c390c5b79108b53ab513943/ktor-client/ktor-client-core/js/src/io/ktor/client/engine/js/JsClientEngine.kt#L62)
+   */
+  // Adding "_capturingHack" to reduce chances of JS IR backend to rename variable,
+  // so it can be accessed inside js("") function
+  @Suppress("UNUSED_PARAMETER", "UnsafeCastFromDynamic", "UNUSED_VARIABLE", "LocalVariableName")
+  private fun createWebSocket(urlString_capturingHack: String, headers: Headers): WebSocket =
+      if (PlatformUtils.IS_NODE) {
+        val ws_capturingHack = js("eval('require')('ws')")
```

Comment:
Note that this only works because we use Ktor for the HTTP engine and Ktor imports `ws`. I'm fine with that but maybe leave a comment?

Reply:
I can either add a comment or add `ws` as a dependency, whichever way make more sense. Adding the dependency will make this independent of ktor being used for the HTTP engine.

Reply:
That works too !